File: scripts/optimize-images.py
from pathlib import Path
from PIL import Image
import json
import os

ROOT = Path('public')
OUT = ROOT / 'optimized'
EXTENSIONS = {'.png', '.jpg', '.jpeg'}
WIDTHS = (640, 960, 1280, 1920)
report = {'scanned': 0, 'converted': 0, 'beforeBytes': 0, 'afterBytes': 0, 'files': []}

for source in ROOT.rglob('*'):
    if not source.is_file() or OUT in source.parents or source.suffix.lower() not in EXTENSIONS:
        continue
    report['scanned'] += 1
    report['beforeBytes'] += source.stat().st_size
    relative = source.relative_to(ROOT)
    target_dir = OUT / relative.parent
    target_dir.mkdir(parents=True, exist_ok=True)
    base = target_dir / f'{source.stem}.webp'
    with Image.open(source) as image:
        width, height = image.size
        has_alpha = 'A' in image.getbands() or image.mode in ('P', 'LA')
        image = image.convert('RGBA' if has_alpha else 'RGB')
        targets = [None, *WIDTHS]
        outputs = []
        for target_width in targets:
            suffix = '' if target_width is None else f'-{target_width}'
            output = target_dir / f'{source.stem}{suffix}.webp'
            if not output.exists() or output.stat().st_mtime < source.stat().st_mtime:
                resized = image if target_width is None or target_width >= width else image.resize((target_width, round(height * target_width / width)), Image.Resampling.LANCZOS)
                resized.save(output, 'WEBP', quality=82, method=6, lossless=False)
            outputs.append(str(output.relative_to(ROOT)).replace('\\', '/'))
        report['converted'] += len(outputs)
        report['afterBytes'] += sum((ROOT / item).stat().st_size for item in outputs)
        report['files'].append({'source': str(relative).replace('\\', '/'), 'width': width, 'height': height, 'outputs': outputs})

(OUT / 'optimization-report.json').write_text(json.dumps(report, ensure_ascii=False, indent=2), encoding='utf-8')
print(json.dumps({k: v for k, v in report.items() if k != 'files'}, ensure_ascii=False, indent=2))

File: scripts/optimize-images.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import sharp from 'sharp';

const root = path.resolve('public');
const rasterExtensions = new Set(['.png', '.jpg', '.jpeg']);
const generatedSuffix = /-(640|960|1280|1920)\.webp$/i;
let scanned = 0;
let converted = 0;
let beforeBytes = 0;
let afterBytes = 0;

async function walk(directory) {
  const entries = await fs.readdir(directory, { withFileTypes: true });
  for (const entry of entries) {
    const filePath = path.join(directory, entry.name);
    if (entry.isDirectory()) {
      if (entry.name !== 'optimized') await walk(filePath);
      continue;
    }
    const extension = path.extname(entry.name).toLowerCase();
    if (!rasterExtensions.has(extension) || generatedSuffix.test(entry.name)) continue;
    await optimize(filePath);
  }
}

async function optimize(filePath) {
  scanned += 1;
  const sourceStat = await fs.stat(filePath);
  beforeBytes += sourceStat.size;
  const metadata = await sharp(filePath).metadata();
  const directory = path.dirname(filePath);
  const baseName = path.basename(filePath, path.extname(filePath));
  const baseOutput = path.join(directory, `${baseName}.webp`);
  const widths = metadata.width && metadata.width > 960 ? [640, 960, 1280, 1920] : [null];

  for (const width of widths) {
    const output = width ? path.join(directory, `${baseName}-${width}.webp`) : baseOutput;
    const outputStat = await fs.stat(output).catch(() => null);
    if (outputStat && outputStat.mtimeMs >= sourceStat.mtimeMs) {
      afterBytes += outputStat.size;
      continue;
    }
    let pipeline = sharp(filePath, { animated: false });
    if (width) pipeline = pipeline.resize({ width, withoutEnlargement: true });
    await pipeline.webp({ quality: 82, alphaQuality: 90, effort: 4 }).toFile(output);
    const resultStat = await fs.stat(output);
    afterBytes += resultStat.size;
    converted += 1;
  }
}

await walk(root);
console.log(JSON.stringify({ scanned, converted, beforeBytes, afterBytes }, null, 2));
